pkg/generators/markers.go

```diff
@@ -0,0 +1,217 @@
+package generators
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"strconv"
+	"strings"
+
+	"k8s.io/gengo/types"
+)
+
+// Subset of spec.SchemaProps that may be placed into a marker comment
+type ValueValidations struct {
+	Maximum          *float64 `json:"maximum,omitempty"`
+	ExclusiveMaximum *bool    `json:"exclusiveMaximum,omitempty"`
+	Minimum          *float64 `json:"minimum,omitempty"`
+	ExclusiveMinimum *bool    `json:"exclusiveMinimum,omitempty"`
+	MaxLength        *int64   `json:"maxLength,omitempty"`
+	MinLength        *int64   `json:"minLength,omitempty"`
+	Pattern          *string  `json:"pattern,omitempty"`
+	MaxItems         *int64   `json:"maxItems,omitempty"`
+	MinItems         *int64   `json:"minItems,omitempty"`
+	UniqueItems      *bool    `json:"uniqueItems,omitempty"`
+	MultipleOf       *float64 `json:"multipleOf,omitempty"`
+	MaxProperties    *int64   `json:"maxProperties,omitempty"`
+	MinProperties    *int64   `json:"minProperties,omitempty"`
+}
+
+type CommentTags struct {
+	ValueValidations
+
+	// Future markers can all be parsed into this centralized struct...
+	// Optional bool `json:"optional,omitempty"`
+	// Default  any  `json:"default,omitempty"`
+}
+
+func ParseCommentTags(comments []string) (CommentTags, error) {
+	// Convert the flat list of comments into a nested map by splitting keys
+	// on colons. For example, the following comments:
+	//  +foo=bar
+	//  +foo:bar=baz
+	//  +foo:bar:baz=qux
+	//  +foo:bar:baz:qux=quux
+	//
+	// Would be converted into the following map:
+	//  map[string]any{
+	//    "foo": map[string]any{
+	//      "bar": map[string]any{
+	//        "baz": map[string]any{
+	//          "qux": "quux",
+	//        },
+	//      },
+	//    },
+	//  }
+	//
+	markers, err := parseMarkers(comments)
+	if err != nil {
+		return CommentTags{}, fmt.Errorf("failed to parse marker comments: %w", err)
+	}
+	nested, err := nestMarkers(markers)
+	if err != nil {
+		return CommentTags{}, fmt.Errorf("invalid marker comments: %w", err)
+	}
+
+	// Parse the map into a CommentTags type by marshalling and unmarshalling
+	// as JSON in leiu of an unstructured converter.
+	out, err := json.Marshal(nested)
+	if err != nil {
+		return CommentTags{}, fmt.Errorf("failed to marshal marker comments: %w", err)
+	}
+
+	var commentTags CommentTags
+	if err = json.Unmarshal(out, &commentTags); err != nil {
+		return CommentTags{}, fmt.Errorf("failed to unmarshal marker comments: %w", err)
+	}
+
+	return commentTags, nil
+}
+
+func parseMarkers(markerComments []string) (map[string]any, error) {
```

Comment:
Don't we already have something like this somewhere else? We already parse a lot of markers from comments, have you looked in `openapi.go`?

Reply:
We use the helpers defined here currently: https://github.com/kubernetes/gengo/blob/master/types/comments.go

But they do not serve this use case well. In KEP-4153 we discussed some CEL ergonomics. So specifically these new marker parsers allow us to interpret the following as you owuld expect:

```
// +k8s:validation:cel[0]:rule="self.pylons >= oldSelf.pylons"
// +k8s:validation:cel[0]:message="YOU MUST CONSTRUCT ADDITIONAL PYLONS"
// +k8s:validation:cel[0]:fieldPath=".pylons"
// +k8s:validation:cel[0]:reason="Invalid"
// +k8s:validation:cel[1]:rule="self.base.owner == 'us'"
// +k8s:validation:cel[1]:message="ALL YOUR BASE ARE BELONG TO US"
// +k8s:validation:cel[1]:fieldPath=".base"
// +k8s:validation:cel[1]:reason="Forbidden"
```

With the existing parser, this would have to be expressed on a single line.